pkg/controller/endpointslicemirroring/endpointslicemirroring_controller.go

```diff
@@ -0,0 +1,493 @@
+// EndpointSlice has changed from or to this controller.
+func (c *Controller) onEndpointSliceUpdate(prevObj, obj interface{}) {
+	prevEndpointSlice := obj.(*discovery.EndpointSlice)
+	endpointSlice := prevObj.(*discovery.EndpointSlice)
+	if endpointSlice == nil || prevEndpointSlice == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointSliceUpdated() expected type discovery.EndpointSlice, got %T, %T", prevObj, obj))
+		return
+	}
+	if managedByChanged(prevEndpointSlice, endpointSlice) || (managedByController(endpointSlice) && c.endpointSliceTracker.Stale(endpointSlice)) {
+		c.queueEndpointsForEndpointSlice(endpointSlice)
+	}
+}
+
+// onEndpointSliceDelete queues a sync for the relevant Endpoints resource for a
+// sync if the EndpointSlice resource version does not match the expected
+// version in the endpointSliceTracker.
+func (c *Controller) onEndpointSliceDelete(obj interface{}) {
+	endpointSlice := getEndpointSliceFromDeleteAction(obj)
+	if endpointSlice == nil {
+		utilruntime.HandleError(fmt.Errorf("onEndpointSliceDelete() expected type discovery.EndpointSlice, got %T", obj))
+		return
+	}
+	if managedByController(endpointSlice) && c.endpointSliceTracker.Has(endpointSlice) {
+		c.queueEndpointsForEndpointSlice(endpointSlice)
+	}
+}
+
+// queueEndpointsForEndpointSlice attempts to queue the corresponding Endpoints
+// resource for the provided EndpointSlice.
+func (c *Controller) queueEndpointsForEndpointSlice(endpointSlice *discovery.EndpointSlice) {
+	key, err := endpointsControllerKey(endpointSlice)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for EndpointSlice %+v (type %T): %v", endpointSlice, endpointSlice, err))
+		return
+	}
+
+	c.queue.AddAfter(key, c.endpointUpdatesBatchPeriod)
+}
+
+// countUnlabeledEndpoints returns the number of Endpoints that are expected to
+// have a skip-mirror label applied by the Endpoints controller but do not. This
+// is computed by counting the number of Services with a selector specified that
+// have a corresponding Endpoints resource without a skip-mirror label set to
+// true.
+func countUnlabeledEndpoints(epLister corelisters.EndpointsLister, svcLister corelisters.ServiceLister) (int, error) {
+	endpoints, err := epLister.List(labels.Everything())
+	if err != nil {
+		return 0, err
+	}
+	services, err := svcLister.List(labels.Everything())
+	if err != nil {
+		return 0, err
+	}
+
+	servicesByNamespacedName := map[string]*v1.Service{}
+	for _, svc := range services {
+		servicesByNamespacedName[namespacedName(svc.Namespace, svc.Name)] = svc
+	}
+
+	numEndpointsWithProblems := 0
+	for _, ep := range endpoints {
+		svc, ok := servicesByNamespacedName[namespacedName(ep.Namespace, ep.Name)]
+		if ok && svc.Spec.Selector != nil {
+			labelVal, ok := ep.Labels[discovery.LabelSkipMirror]
+			if !ok || labelVal != "true" {
+				klog.V(4).Infof("Expected %s Endpoints resource to have %s label set to true, got %s", ep.Name, discovery.LabelSkipMirror, labelVal)
```

Comment:
This is a candidate for V(5) if any

Reply:
Moved to v(5). I think this is still valuable as this line can give insight into why the EndpointSliceMirroring controller isn't starting up if that ever happens. Without it, we just have a number of Endpoints, while this gives us actual names of problematic Endpoints. I agree that this could be unnecessarily verbose in most normal startup cases.

Reply:
This is now irrelevant (and removed) thanks to Minhan's suggestion to not require the label to be set here.